File: backend/src/controllers/incomeController.ts
import { Request, Response, NextFunction } from 'express';
import incomeService from '../services/incomeService';
import { ApiError } from '../utils/errorHandler';
import { AuthRequest } from '../middleware/auth';
import { IncomeFilterParams } from '../types/income';

/**
 * Create income
 * POST /api/income
 */
export const createIncome = async (
  req: AuthRequest,
  res: Response,
  next: NextFunction
) => {
  try {
    const userId = req.userId;

    if (!userId) {
      throw new ApiError(401, 'User not authenticated');
    }

    const result = await incomeService.createIncome(req.body, userId);

    res.status(201).json({
      success: true,
      message: 'Income created successfully',
      data: result,
    });
  } catch (error) {
    next(error);
  }
};

/**
 * Get income by ID
 * GET /api/income/:id
 */
export const getIncome = async (
  req: Request,
  res: Response,
  next: NextFunction
) => {
  try {
    const { id } = req.params;

    const result = await incomeService.getIncomeById(id);

    res.status(200).json({
      success: true,
      data: result,
    });
  } catch (error) {
    next(error);
  }
};

/**
 * Get all income with filters
 * GET /api/income
 */
export const getIncomes = async (
  req: Request,
  res: Response,
  next: NextFunction
) => {
  try {
    const filters: IncomeFilterParams = {
      page: req.query.page ? parseInt(req.query.page as string) : 1,
      pageSize: req.query.pageSize ? parseInt(req.query.pageSize as string) : 10,
      startDate: req.query.startDate as string,
      endDate: req.query.endDate as string,
      categoryId: req.query.categoryId as string,
      search: req.query.search as string,
      sortBy: (req.query.sortBy as any) || 'date',
      sortOrder: (req.query.sortOrder as any) || 'desc',
    };

    const result = await incomeService.getIncomes(filters);

    res.status(200).json({
      success: true,
      data: result,
    });
  } catch (error) {
    next(error);
  }
};

/**
 * Update income
 * PUT /api/income/:id
 */
export const updateIncome = async (
  req: AuthRequest,
  res: Response,
  next: NextFunction
) => {
  try {
    const { id } = req.params;
    const userId = req.userId;

    if (!userId) {
      throw new ApiError(401, 'User not authenticated');
    }

    const result = await incomeService.updateIncome(id, req.body, userId);

    res.status(200).json({
      success: true,
      message: 'Income updated successfully',
      data: result,
    });
  } catch (error) {
    next(error);
  }
};

/**
 * Delete income
 * DELETE /api/income/:id
 */
export const deleteIncome = async (
  req: AuthRequest,
  res: Response,
  next: NextFunction
) => {
  try {
    const { id } = req.params;
    const userId = req.userId;

    if (!userId) {
      throw new ApiError(401, 'User not authenticated');
    }

    await incomeService.deleteIncome(id, userId);

    res.status(200).json({
      success: true,
      message: 'Income deleted successfully',
    });
  } catch (error) {
    next(error);
  }
};

/**
 * Get income summary
 * GET /api/income/summary
 */
export const getIncomeSummary = async (
  req: Request,
  res: Response,
  next: NextFunction
) => {
  try {
    const { startDate, endDate } = req.query;

    const result = await incomeService.getIncomeSummary(
      startDate as string,
      endDate as string
    );

    res.status(200).json({
      success: true,
      data: result,
    });
  } catch (error) {
    next(error);
  }
};

File: backend/MadrasaAccounting.API/Controllers/IncomeController.cs
using MadrasaAccounting.Application.DTOs.Income;
using MadrasaAccounting.Application.Services;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;

namespace MadrasaAccounting.API.Controllers;

[ApiController]
[Route("api/[controller]")]
[Authorize]
public class IncomeController : ControllerBase
{
    private readonly IIncomeService _incomeService;
    private readonly ILogger<IncomeController> _logger;

    public IncomeController(IIncomeService incomeService, ILogger<IncomeController> logger)
    {
        _incomeService = incomeService;
        _logger = logger;
    }

    /// <summary>
    /// Create new income record
    /// </summary>
    [HttpPost]
    [ProducesResponseType(StatusCodes.Status201Created)]
    [ProducesResponseType(StatusCodes.Status400BadRequest)]
    public async Task<IActionResult> CreateIncome([FromBody] CreateIncomeRequest request)
    {
        try
        {
            var adminIdClaim = User.FindFirst("sub");
            if (adminIdClaim == null || !Guid.TryParse(adminIdClaim.Value, out var adminId))
            {
                return Unauthorized();
            }

            var income = await _incomeService.CreateIncomeAsync(adminId, request);

            return CreatedAtAction(nameof(GetIncome), new { id = income.Id }, new
            {
                success = true,
                data = income,
                message = "Income created successfully",
                timestamp = DateTime.UtcNow
            });
        }
        catch (InvalidOperationException ex)
        {
            return BadRequest(new
            {
                success = false,
                message = ex.Message,
                errors = new[] { ex.Message },
                timestamp = DateTime.UtcNow
            });
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error creating income");
            return StatusCode(StatusCodes.Status500InternalServerError, new
            {
                success = false,
                message = "An error occurred while creating income",
                errors = new string[0],
                timestamp = DateTime.UtcNow
            });
        }
    }

    /// <summary>
    /// Get income by ID
    /// </summary>
    [HttpGet("{id}")]
    [ProducesResponseType(StatusCodes.Status200OK)]
    [ProducesResponseType(StatusCodes.Status404NotFound)]
    public async Task<IActionResult> GetIncome(Guid id)
    {
        try
        {
            var adminIdClaim = User.FindFirst("sub");
            if (adminIdClaim == null || !Guid.TryParse(adminIdClaim.Value, out var adminId))
            {
                return Unauthorized();
            }

            var income = await _incomeService.GetIncomeByIdAsync(adminId, id);

            return Ok(new
            {
                success = true,
                data = income,
                message = "Income retrieved successfully",
                timestamp = DateTime.UtcNow
            });
        }
        catch (InvalidOperationException)
        {
            return NotFound(new
            {
                success = false,
                message = "Income not found",
                errors = new[] { "Income not found" },
                timestamp = DateTime.UtcNow
            });
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error retrieving income");
            return StatusCode(StatusCodes.Status500InternalServerError, new
            {
                success = false,
                message = "An error occurred while retrieving income",
                errors = new string[0],
                timestamp = DateTime.UtcNow
            });
        }
    }

    /// <summary>
    /// Get paginated income list with date filtering
    /// </summary>
    [HttpGet]
    [ProducesResponseType(StatusCodes.Status200OK)]
    public async Task<IActionResult> GetIncomeList(
        [FromQuery] int pageNumber = 1,
        [FromQuery] int pageSize = 10,
        [FromQuery] DateTime? startDate = null,
        [FromQuery] DateTime? endDate = null)
    {
        try
        {
            var adminIdClaim = User.FindFirst("sub");
            if (adminIdClaim == null || !Guid.TryParse(adminIdClaim.Value, out var adminId))
            {
                return Unauthorized();
            }

            var result = await _incomeService.GetIncomeListAsync(adminId, pageNumber, pageSize, startDate, endDate);

            return Ok(new
            {
                success = true,
                data = result,
                message = "Income list retrieved successfully",
                timestamp = DateTime.UtcNow
            });
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error retrieving income list");
            return StatusCode(StatusCodes.Status500InternalServerError, new
            {
                success = false,
                message = "An error occurred while retrieving income list",
                errors = new string[0],
                timestamp = DateTime.UtcNow
            });
        }
    }

    /// <summary>
    /// Search income by description or category
    /// </summary>
    [HttpGet("search/{searchTerm}")]
    [ProducesResponseType(StatusCodes.Status200OK)]
    public async Task<IActionResult> SearchIncome(
        string searchTerm,
        [FromQuery] int pageNumber = 1,
        [FromQuery] int pageSize = 10)
    {
        try
        {
            var adminIdClaim = User.FindFirst("sub");
            if (adminIdClaim == null || !Guid.TryParse(adminIdClaim.Value, out var adminId))
            {
                return Unauthorized();
            }

            var result = await _incomeService.SearchIncomeAsync(adminId, searchTerm, pageNumber, pageSize);

            return Ok(new
            {
                success = true,
                data = result,
                message = "Income search results retrieved successfully",
                timestamp = DateTime.UtcNow
            });
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error searching income");
            return StatusCode(StatusCodes.Status500InternalServerError, new
            {
                success = false,
                message = "An error occurred while searching income",
                errors = new string[0],
                timestamp = DateTime.UtcNow
            });
        }
    }

    /// <summary>
    /// Update income record
    /// </summary>
    [HttpPut("{id}")]
    [ProducesResponseType(StatusCodes.Status200OK)]
    [ProducesResponseType(StatusCodes.Status400BadRequest)]
    [ProducesResponseType(StatusCodes.Status404NotFound)]
    public async Task<IActionResult> UpdateIncome(Guid id, [FromBody] UpdateIncomeRequest request)
    {
        try
        {
            var adminIdClaim = User.FindFirst("sub");
            if (adminIdClaim == null || !Guid.TryParse(adminIdClaim.Value, out var adminId))
            {
                return Unauthorized();
            }

            var income = await _incomeService.UpdateIncomeAsync(adminId, id, request);

            return Ok(new
            {
                success = true,
                data = income,
                message = "Income updated successfully",
                timestamp = DateTime.UtcNow
            });
        }
        catch (InvalidOperationException ex)
        {
            return ex.Message.Contains("not found")
                ? NotFound(new
                {
                    success = false,
                    message = ex.Message,
                    errors = new[] { ex.Message },
                    timestamp = DateTime.UtcNow
                })
                : BadRequest(new
                {
                    success = false,
                    message = ex.Message,
                    errors = new[] { ex.Message },
                    timestamp = DateTime.UtcNow
                });
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error updating income");
            return StatusCode(StatusCodes.Status500InternalServerError, new
            {
                success = false,
                message = "An error occurred while updating income",
                errors = new string[0],
                timestamp = DateTime.UtcNow
            });
        }
    }

    /// <summary>
    /// Delete income record
    /// </summary>
    [HttpDelete("{id}")]
    [ProducesResponseType(StatusCodes.Status200OK)]
    [ProducesResponseType(StatusCodes.Status404NotFound)]
    public async Task<IActionResult> DeleteIncome(Guid id)
    {
        try
        {
            var adminIdClaim = User.FindFirst("sub");
            if (adminIdClaim == null || !Guid.TryParse(adminIdClaim.Value, out var adminId))
            {
                return Unauthorized();
            }

            await _incomeService.DeleteIncomeAsync(adminId, id);

            return Ok(new
            {
                success = true,
                message = "Income deleted successfully",
                timestamp = DateTime.UtcNow
            });
        }
        catch (InvalidOperationException)
        {
            return NotFound(new
            {
                success = false,
                message = "Income not found",
                errors = new[] { "Income not found" },
                timestamp = DateTime.UtcNow
            });
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error deleting income");
            return StatusCode(StatusCodes.Status500InternalServerError, new
            {
                success = false,
                message = "An error occurred while deleting income",
                errors = new string[0],
                timestamp = DateTime.UtcNow
            });
        }
    }
}
